### PR TITLE
feat: Full support for quantifiers in Python

### PR DESCRIPTION
**Pull request by fabiomadge**
Builds on #2646.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 1]
```diff
@@ -989,15 +995,17 @@ protected class ClassWriter : IClassWriter {
...
 
     protected override void EmitEmptyTupleList(string tupleTypeArgs, ConcreteSyntaxTree wr) {
-      throw new UnsupportedFeatureException(Token.NoToken, Feature.NonSequentializableForallStatements);
+      wr.Write("[]");
     }
 
     protected override ConcreteSyntaxTree EmitAddTupleToList(string ingredients, string tupleTypeArgs, ConcreteSyntaxTree wr) {
-      throw new UnsupportedFeatureException(Token.NoToken, Feature.NonSequentializableForallStatements);
+      var wrTuple = new ConcreteSyntaxTree();
+      wr.FormatLine($"{ingredients}.append(({wrTuple}))");
```
Comment (cpitclaudel) [inline Source/Dafny/Compilers/Compiler-python.cs:1003]: How does this part work?  It looks like you're serializing `wrTuple` to string before it's been filled in?

**Review by cpitclaudel — DISMISSED** [review]

File: Source/Dafny/Compilers/Compiler-python.cs (hunk 1 above)
Comment (fabiomadge) [reply Source/Dafny/Compilers/Compiler-python.cs:1003]: Yes, but I'm doing `FormatLine`, which will be evaluated when serialising the `ConcreteSyntaxTree`.

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 2]
```diff
@@ -1584,29 +1599,43 @@ protected class ClassWriter : IClassWriter {
...
+      string typeTest;
       if (boundVarType.IsRefType) {
-        throw new UnsupportedFeatureException(tok, Feature.SubtypeConstraintsInQuantifiers);
+        if (boundVarType.IsObject || boundVarType.IsObjectQ) {
+          typeTest = True;
+        } else {
+          typeTest = $"isinstance({tmpVarName}, {TypeName(boundVarType, wPreconditions, tok)})";
+        }
+        if (boundVarType.IsNonNullRefType) {
+          typeTest = $"{tmpVarName} is not None and {typeTest}";
```
[the remark this answers is not in the document]
Comment (fabiomadge) [reply Source/Dafny/Compilers/Compiler-python.cs:1610]: The `typeTest` is guaranteed to be a closed expression and the rest seems to be fine.

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 3]
```diff
@@ -1584,29 +1599,43 @@ protected class ClassWriter : IClassWriter {
 
     [CanBeNull]
     protected override string GetSubtypeCondition(string tmpVarName, Type boundVarType, IToken tok, ConcreteSyntaxTree wPreconditions) {
+      string typeTest;
       if (boundVarType.IsRefType) {
-        throw new UnsupportedFeatureException(tok, Feature.SubtypeConstraintsInQuantifiers);
+        if (boundVarType.IsObject || boundVarType.IsObjectQ) {
+          typeTest = True;
```
[the remark this answers is not in the document]
Comment (fabiomadge) [reply Source/Dafny/Compilers/Compiler-python.cs:1605]: This function is strange. We add this test a couple lines later.

**Review by cpitclaudel — APPROVED** [review]

File: Source/DafnyRuntime/DafnyRuntime.py [hunk 4]
```diff
@@ -341,6 +361,26 @@ def quantifier(vals, frall, pred):
...
+    if lo is None:
+        return count(hi-1, -1)
+    elif hi is None:
+        return count(lo)
+    return range(lo, hi)
+
+class Doubler:
+    def __init__(self, start):
+        self.start = start
+    def __iter__(self):
```
[the remark this answers is not in the document]
Comment (cpitclaudel) [reply Source/DafnyRuntime/DafnyRuntime.py:377]: Yes, but you had to call `iter` on the object first: each call to `iter` generates a fresh iterator.  The `Doubler` here won't behave like that.

File: Source/Dafny/Compilers/Compiler-python.cs (hunk 1 above)
Comment (cpitclaudel) [reply Source/Dafny/Compilers/Compiler-python.cs:1003]: TIL that `$"…"` is lazy sometimes.

File: Source/DafnyRuntime/DafnyRuntime.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabiomadge) [reply Source/DafnyRuntime/DafnyRuntime.py:377]: Good point. Addressed in #2663. As you noted, we never give those `Doubler`s a name, so it's not a problem for us, but I fixed it along with an off-by-one.

**Pull request merged**